File: client/android/app/src/main/kotlin/com/example/shenliyuan/GradeReminderSnapshot.kt
package com.example.shenliyuan

import org.json.JSONArray
import org.json.JSONObject
import java.security.MessageDigest

data class GradeReminderSnapshot(
    val initialized: Boolean,
    val grades: List<GradeReminderItem>,
    val updatedAt: Long,
) {
    fun toJson(): JSONObject {
        val array = JSONArray()
        grades.forEach { array.put(it.toJson()) }
        return JSONObject()
            .put("initialized", initialized)
            .put("updatedAt", updatedAt)
            .put("grades", array)
    }

    fun diffFrom(old: GradeReminderSnapshot?): GradeReminderDiff {
        if (old == null || !old.initialized) {
            return GradeReminderDiff(emptyList(), baselineOnly = true)
        }

        val oldByKey = old.grades.associateBy { it.key }
        val changes = mutableListOf<GradeReminderChange>()

        grades.forEach { next ->
            val previous = oldByKey[next.key]
            if (previous == null) {
                if (next.hasEffectiveGrade()) {
                    changes += GradeReminderChange(null, next, "added")
                }
            } else if (
                (!previous.hasEffectiveGrade() && next.hasEffectiveGrade()) ||
                previous.grade.trim() != next.grade.trim() ||
                previous.gpa.orEmpty() != next.gpa.orEmpty() ||
                previous.fraction.orEmpty() != next.fraction.orEmpty()
            ) {
                if (
                    previous.hasEffectiveGrade() ||
                    next.hasEffectiveGrade() ||
                    !previous.gpa.isNullOrBlank() ||
                    !next.gpa.isNullOrBlank() ||
                    !previous.fraction.isNullOrBlank() ||
                    !next.fraction.isNullOrBlank()
                ) {
                    changes += GradeReminderChange(previous, next, "updated")
                }
            }
        }

        return GradeReminderDiff(changes.sortedBy { it.next.name })
    }

    companion object {
        fun empty(): GradeReminderSnapshot =
            GradeReminderSnapshot(true, emptyList(), System.currentTimeMillis())

        fun fromGrades(array: JSONArray): GradeReminderSnapshot {
            val items = buildList {
                for (index in 0 until array.length()) {
                    val item = fromGradeJson(array.optJSONObject(index) ?: continue)
                    if (item.key.isNotBlank()) add(item)
                }
            }.sortedBy { it.key }
            return GradeReminderSnapshot(true, items, System.currentTimeMillis())
        }

        fun fromJson(raw: String?): GradeReminderSnapshot? {
            if (raw.isNullOrBlank()) return null
            return runCatching {
                val json = JSONObject(raw)
                val array = json.optJSONArray("grades") ?: JSONArray()
                val items = buildList {
                    for (index in 0 until array.length()) {
                        GradeReminderItem.fromJson(array.optJSONObject(index) ?: continue)
                            ?.let(::add)
                    }
                }
                GradeReminderSnapshot(
                    initialized = json.optBoolean("initialized", false),
                    grades = items,
                    updatedAt = json.optLong("updatedAt", System.currentTimeMillis()),
                )
            }.getOrNull()
        }

        fun fromJsonObject(json: JSONObject?): GradeReminderSnapshot? {
            if (json == null) return null
            return fromJson(json.toString())
        }

        private fun fromGradeJson(json: JSONObject): GradeReminderItem {
            val name = json.optStringOrNull("name").orEmpty()
            val credits = normalizeNumber(json.opt("credits"))
            val examType = blankToNull(json.optStringOrNull("exam_type"))
            val key = stableKey(json, name, credits, examType)
            return GradeReminderItem(
                key = key,
                name = name,
                grade = json.optStringOrNull("grade") ?: "--",
                gpa = blankToNull(normalizeNumber(json.opt("gpa"))),
                fraction = blankToNull(normalizeNumber(json.opt("fraction"))),
                credits = credits.ifBlank { "0" },
                examType = examType,
                isDegree = json.optBoolean("is_degree", false) ||
                    json.optStringOrNull("is_degree") == "1" ||
                    json.optStringOrNull("is_degree") == "是",
            )
        }

        private fun stableKey(
            json: JSONObject,
            name: String,
            credits: String,
            examType: String?,
        ): String {
            val studentGradeId = json.optStringOrNull("student_grade_id").orEmpty()
            if (studentGradeId.isNotEmpty()) return "student:$studentGradeId"
            val classId = json.optStringOrNull("class_id").orEmpty()
            if (classId.isNotEmpty()) return "class:$classId"
            val courseId = json.optStringOrNull("course_id").orEmpty()
            if (courseId.isNotEmpty()) return "course:$courseId"
            val courseCode = json.optStringOrNull("course_code").orEmpty()
            if (courseCode.isNotEmpty() && name.isNotEmpty()) {
                return "code:$courseCode|$name"
            }
            if (name.isNotEmpty()) return "name:$name|$credits|${examType.orEmpty()}"
            return ""
        }

        private fun normalizeNumber(value: Any?): String {
            val text = value?.toString()?.trim().orEmpty()
            if (text.isBlank() || text == "null") return ""
            val asDouble = text.toDoubleOrNull() ?: return text
            val asLong = asDouble.toLong()
            return if (asDouble == asLong.toDouble()) asLong.toString() else text
        }

        private fun blankToNull(value: String?): String? {
            val text = value?.trim().orEmpty()
            return text.ifBlank { null }
        }
    }
}

data class GradeReminderItem(
    val key: String,
    val name: String,
    val grade: String,
    val gpa: String?,
    val fraction: String?,
    val credits: String,
    val examType: String?,
    val isDegree: Boolean,
) {
    fun hasEffectiveGrade(): Boolean {
        val text = grade.trim()
        if (text.isBlank()) return false
        return text !in setOf("--", "未录入", "暂无", "无", "null", "NULL", "缓考")
    }

    fun displayGrade(): String = grade.trim().ifBlank { "--" }

    fun toJson(): JSONObject = JSONObject()
        .put("key", key)
        .put("name", name)
        .put("grade", grade)
        .put("gpa", gpa)
        .put("fraction", fraction)
        .put("credits", credits)
        .put("examType", examType)
        .put("isDegree", isDegree)

    companion object {
        fun fromJson(json: JSONObject): GradeReminderItem? {
            val key = json.optString("key")
            if (key.isBlank()) return null
            return GradeReminderItem(
                key = key,
                name = json.optString("name"),
                grade = json.optString("grade"),
                gpa = json.optStringOrNull("gpa"),
                fraction = json.optStringOrNull("fraction"),
                credits = json.optString("credits", "0"),
                examType = json.optStringOrNull("examType"),
                isDegree = json.optBoolean("isDegree", false),
            )
        }
    }
}

data class GradeReminderChange(
    val previous: GradeReminderItem?,
    val next: GradeReminderItem,
    val type: String,
) {
    fun summary(): String = "${next.name}: ${previous?.displayGrade() ?: "--"} -> ${next.displayGrade()}"

    fun toHashJson(): JSONObject = JSONObject()
        .put("type", type)
        .put("key", next.key)
        .put("name", next.name)
        .put("oldGrade", previous?.displayGrade().orEmpty())
        .put("newGrade", next.displayGrade())
        .put("oldGpa", previous?.gpa.orEmpty())
        .put("newGpa", next.gpa.orEmpty())
        .put("oldFraction", previous?.fraction.orEmpty())
        .put("newFraction", next.fraction.orEmpty())
}

data class GradeReminderDiff(
    val changes: List<GradeReminderChange>,
    val baselineOnly: Boolean = false,
) {
    val hasChanges: Boolean get() = changes.isNotEmpty()

    fun changeHash(): String {
        if (changes.isEmpty()) return ""
        val array = JSONArray()
        changes.forEach { array.put(it.toHashJson()) }
        return sha1(array.toString())
    }

    private fun sha1(value: String): String {
        val digest = MessageDigest.getInstance("SHA-1")
            .digest(value.toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }
}

private fun JSONObject.optStringOrNull(name: String): String? {
    if (!has(name) || isNull(name)) return null
    val text = optString(name).trim()
    return text.ifBlank { null }
}

File: client/lib/models/grade_reminder_snapshot.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';

import 'edu_grade.dart';

class GradeReminderSnapshot {
  final bool initialized;
  final List<GradeReminderSnapshotItem> grades;
  final DateTime updatedAt;

  const GradeReminderSnapshot({
    required this.initialized,
    required this.grades,
    required this.updatedAt,
  });

  factory GradeReminderSnapshot.empty({DateTime? updatedAt}) {
    return GradeReminderSnapshot(
      initialized: true,
      grades: const [],
      updatedAt: updatedAt ?? DateTime.now(),
    );
  }

  factory GradeReminderSnapshot.fromGrades(
    List<EduGrade> grades, {
    DateTime? updatedAt,
  }) {
    final items = grades
        .map(GradeReminderSnapshotItem.fromGrade)
        .where((item) => item.key.isNotEmpty)
        .toList()
      ..sort((a, b) => a.key.compareTo(b.key));

    return GradeReminderSnapshot(
      initialized: true,
      grades: items,
      updatedAt: updatedAt ?? DateTime.now(),
    );
  }

  factory GradeReminderSnapshot.initialBaseline(
    List<EduGrade> grades, {
    DateTime? updatedAt,
  }) {
    final snapshot = GradeReminderSnapshot.fromGrades(
      grades,
      updatedAt: updatedAt,
    );
    if (snapshot.grades.isNotEmpty) return snapshot;
    return GradeReminderSnapshot(
      initialized: false,
      grades: const [],
      updatedAt: snapshot.updatedAt,
    );
  }

  factory GradeReminderSnapshot.fromJson(Map<String, dynamic> json) {
    final rawGrades = json['grades'];
    return GradeReminderSnapshot(
      initialized: json['initialized'] == true,
      grades: rawGrades is List
          ? rawGrades
              .whereType<Map>()
              .map((item) => GradeReminderSnapshotItem.fromJson(
                    Map<String, dynamic>.from(item),
                  ))
              .where((item) => item.key.isNotEmpty)
              .toList()
          : const [],
      updatedAt: DateTime.fromMillisecondsSinceEpoch(
        (json['updatedAt'] as num?)?.toInt() ??
            DateTime.now().millisecondsSinceEpoch,
      ),
    );
  }

  Map<String, dynamic> toJson() => {
        'initialized': initialized,
        'updatedAt': updatedAt.millisecondsSinceEpoch,
        'grades': grades.map((item) => item.toJson()).toList(),
      };

  GradeReminderDiff diffFrom(GradeReminderSnapshot? oldSnapshot) {
    if (oldSnapshot == null ||
        !oldSnapshot.initialized ||
        oldSnapshot.grades.isEmpty) {
      return const GradeReminderDiff(changes: [], baselineOnly: true);
    }

    final oldByKey = {
      for (final item in oldSnapshot.grades) item.key: item,
    };
    final changes = <GradeReminderChange>[];

    for (final next in grades) {
      final old = oldByKey[next.key];
      if (old == null) {
        if (next.hasEffectiveGrade) {
          changes.add(GradeReminderChange.added(next));
        }
        continue;
      }

      if (_gradeBecameEffective(old, next) ||
          _changed(old.grade, next.grade) ||
          _changed(old.gpa, next.gpa) ||
          _changed(old.fraction, next.fraction)) {
        if (next.hasEffectiveGrade ||
            old.hasEffectiveGrade ||
            next.gpa != null ||
            next.fraction != null) {
          changes.add(GradeReminderChange.updated(old, next));
        }
      }
    }

    changes.sort((a, b) => a.next.name.compareTo(b.next.name));
    return GradeReminderDiff(changes: changes);
  }

  static bool _gradeBecameEffective(
    GradeReminderSnapshotItem old,
    GradeReminderSnapshotItem next,
  ) {
    return !old.hasEffectiveGrade && next.hasEffectiveGrade;
  }

  static bool _changed(String? a, String? b) {
    return (a ?? '').trim() != (b ?? '').trim();
  }
}

class GradeReminderSnapshotItem {
  final String key;
  final String name;
  final String grade;
  final String? gpa;
  final String? fraction;
  final String credits;
  final String? examType;
  final bool isDegree;

  const GradeReminderSnapshotItem({
    required this.key,
    required this.name,
    required this.grade,
    required this.gpa,
    required this.fraction,
    required this.credits,
    required this.examType,
    required this.isDegree,
  });

  factory GradeReminderSnapshotItem.fromGrade(EduGrade grade) {
    final name = grade.name.trim();
    final credits = _formatNumber(grade.credits);
    final examType = _blankToNull(grade.examType);
    return GradeReminderSnapshotItem(
      key: _stableKey(grade, name, credits, examType),
      name: name,
      grade: grade.displayGrade.trim(),
      gpa: _formatNullableNumber(grade.gpa),
      fraction: _formatNullableNumber(grade.fraction),
      credits: credits,
      examType: examType,
      isDegree: grade.isDegree,
    );
  }

  factory GradeReminderSnapshotItem.fromJson(Map<String, dynamic> json) {
    return GradeReminderSnapshotItem(
      key: json['key']?.toString() ?? '',
      name: json['name']?.toString() ?? '',
      grade: json['grade']?.toString() ?? '',
      gpa: _blankToNull(json['gpa']),
      fraction: _blankToNull(json['fraction']),
      credits: json['credits']?.toString() ?? '0',
      examType: _blankToNull(json['examType'] ?? json['exam_type']),
      isDegree: json['isDegree'] == true || json['is_degree'] == true,
    );
  }

  bool get hasEffectiveGrade {
    final text = grade.trim();
    if (text.isEmpty) return false;
    const placeholders = {
      '--',
      '未录入',
      '暂无',
      '无',
      'null',
      'NULL',
      '缓考',
    };
    return !placeholders.contains(text);
  }

  String get displayGrade => grade.trim().isEmpty ? '--' : grade.trim();

  Map<String, dynamic> toJson() => {
        'key': key,
        'name': name,
        'grade': grade,
        'gpa': gpa,
        'fraction': fraction,
        'credits': credits,
        'examType': examType,
        'isDegree': isDegree,
      };

  static String _stableKey(
    EduGrade grade,
    String name,
    String credits,
    String? examType,
  ) {
    final studentGradeId = grade.studentGradeId.trim();
    if (studentGradeId.isNotEmpty) return 'student:$studentGradeId';
    final classId = grade.classId.trim();
    if (classId.isNotEmpty) return 'class:$classId';
    final courseId = grade.courseId.trim();
    if (courseId.isNotEmpty) return 'course:$courseId';
    final courseCode = grade.courseCode.trim();
    if (courseCode.isNotEmpty && name.isNotEmpty) {
      return 'code:$courseCode|$name';
    }
    if (name.isNotEmpty) {
      return 'name:$name|$credits|${examType ?? ''}';
    }
    return '';
  }

  static String _formatNumber(double value) {
    if (value == value.roundToDouble()) {
      return value.toInt().toString();
    }
    return value.toString();
  }

  static String? _formatNullableNumber(double? value) {
    if (value == null) return null;
    return _formatNumber(value);
  }

  static String? _blankToNull(dynamic value) {
    if (value == null) return null;
    final text = value.toString().trim();
    return text.isEmpty ? null : text;
  }
}

enum GradeReminderChangeType { added, updated }

class GradeReminderChange {
  final GradeReminderChangeType type;
  final GradeReminderSnapshotItem? previous;
  final GradeReminderSnapshotItem next;

  const GradeReminderChange._({
    required this.type,
    required this.previous,
    required this.next,
  });

  factory GradeReminderChange.added(GradeReminderSnapshotItem next) {
    return GradeReminderChange._(
      type: GradeReminderChangeType.added,
      previous: null,
      next: next,
    );
  }

  factory GradeReminderChange.updated(
    GradeReminderSnapshotItem previous,
    GradeReminderSnapshotItem next,
  ) {
    return GradeReminderChange._(
      type: GradeReminderChangeType.updated,
      previous: previous,
      next: next,
    );
  }

  String get summary {
    final oldGrade = previous?.displayGrade ?? '--';
    return '${next.name}: $oldGrade -> ${next.displayGrade}';
  }

  Map<String, dynamic> toHashJson() => {
        'type': type.name,
        'key': next.key,
        'name': next.name,
        'oldGrade': previous?.displayGrade ?? '',
        'newGrade': next.displayGrade,
        'oldGpa': previous?.gpa ?? '',
        'newGpa': next.gpa ?? '',
        'oldFraction': previous?.fraction ?? '',
        'newFraction': next.fraction ?? '',
      };
}

class GradeReminderDiff {
  final List<GradeReminderChange> changes;
  final bool baselineOnly;

  const GradeReminderDiff({
    required this.changes,
    this.baselineOnly = false,
  });

  bool get hasChanges => changes.isNotEmpty;

  String get changeHash {
    if (changes.isEmpty) return '';
    final payload = jsonEncode(changes.map((c) => c.toHashJson()).toList());
    return sha1.convert(utf8.encode(payload)).toString();
  }
}
